### ISSUE TITLE
Evaluation Results Differ from Claimed Results in Paper

### ISSUE THREAD

**Issue by qw85639229** [HelmholtzAI-FZJ/flex_gen#2]
Hi,

I evaluated the checkpoints `GSQ-F16-D16-G8-V256k` downloaded from Huggingface. The results are as follows:
```
!FLAG None |mse:0.0069|fid:1.2969|is_mean:185.0000|is_std:8.8882|lpips:0.1065|psnr:21.5942|ssim:0.5756|perplexity:278.4122|used_codebook:0.3384|prod_perplexity:155.5554|prod_used_codebook:0.0000|
```
These results differ from the rFID=0.50 claimed in the paper.

The run_tokenizer_eval.sh settings were as follows:
```
resolution=256
SAMPLES_FOR_TEST=50000
BATCH_SIZE=32
```
Could you please help clarify the discrepancy?

Thank you.

**Comment by rakkit** [comment]
Hi, there are three versions of F16-D16-G8
```GSQ-F16-D16-G8-V256k``` fid:1.2891
```GSQ-F16-D16-G8-V256k-no_l2```fid:0.5078
```GSQ-F16-D16-G8-V256k-no_l2-no_sphere``` fid:7.0938

We hint in paper in the case of D/G=1 or D/G=2, we don't use L2-normalization.  We are writing a new paper explaining this phenomenon; in short, 1(2)d sphere has very limited space.

**Comment by qw85639229** [comment]
Hello, thank you very much for your response. We have evaluated the results of `GSQ-F16-D16-G8-V256k-no_l2`, which are very close to the results claimed in the paper. However, after training, we were not able to reproduce the corresponding results. Could you please provide the training config.yaml file for `GSQ-F16-D16-G8-V256k-no_l2`?

**Comment by rakkit** [comment]
```
seed: 42
dataset: imagenet-1k
train_batch_size: 4 # per GPU # total should be 512
accelerate:
  mixed_precision: bf16
ema:
  beta: 0.999
model:
  name: flexTokenizer
  embedding_dim: 16
  hidden_channels: 128
  channel_multipliers:
  - 1
  - 1
  - 2
  - 2
  - 4
  encoder_layer_configs:
  - 2
  - 2
  - 2
  - 2
  - 2
  - 2
  decoder_layer_configs:
  - 2
  - 2
  - 2
  - 2
  - 2
  - 2
  use_adaptive_norm: true
  use_learnable_up_down_sample: true
  quantizer_config:
    quantize_type: gvq
    embed_dim: 16
    num_embed: 262144
    commitment_loss_weight: 0.25
    use_l2_norm: false
    use_uniform_init: false
    num_group: 8
discriminator:
  name: dino_discriminator
  lambda_perceptual_loss: 1.0
  disable_adaptive_weight: true
  gen_loss_type: non_saturate
  disc_loss_type: hinge
  lambda_adversarial_loss: 0.1
  lambda_disc_loss: 1.0
logger:
  dataset_name: ${dataset}
  checkpoint_every_step: 5000
  wandb:
    project: any_tokenizer_official
    mode: online
  image_logger:
    img_log_iter_frequency: 1
    max_images_to_log: 8
    save_on_local: true
    save_on_wandb: true
    rescale_mean:
    - 0.5
    - 0.5
    - 0.5
    rescale_std:
    - 0.5
    - 0.5
    - 0.5
  output_dir: ./logs_offical_run_tokenizer/ablation_size/GVQ-D-GAN-256-f16-d16g8-v262144-l2-BS256-1_1_2_2_4-no_l2
train:
  num_train_steps: 50000
  grad_accum_every: 1
  apply_gradient_penalty_every: 0
  max_grad_norm: 2
  discr_start_after_step: 0
  train_batch_size: ${train_batch_size}
evaluation:
  eval_every_step: 2500
  eval_for_steps: 97
  dataset_name: ${dataset}
  metrics:
  - mse
  - fid
  - is
  - lpips
  - psnr
  - ssim
optim:
  base_lr: 0.0002
  lr_scale: fixed
  wd: 0.05
  betas:
  - 0.9
  - 0.99
  eps: 1.0e-08
disc_optim:
  base_lr: 0.0002
  lr_scale: fixed
  wd: 0.05
  betas:
  - 0.9
  - 0.99
  eps: 1.0e-08
lr_scheduler:
  name: constant
disc_lr_scheduler:
  name: constant
train_data:
  dataset:
    path: ~/HF/imagenet-1k
    split: train
  transforms:
    no_aug: true
    is_training: true
    input_size: 256
    mean:
    - 0.5
    - 0.5
    - 0.5
    std:
    - 0.5
    - 0.5
    - 0.5
  dataloader:
    persistent_workers: true
    batch_size: ${train_batch_size}
    num_workers: 20
    shuffle: true
eval_data:
  dataset:
    path: ~/HF/imagenet-1k
    split: validation
  transforms:
    no_aug: true
    is_training: false
    input_size: 256
    mean:
    - 0.5
    - 0.5
    - 0.5
    std:
    - 0.5
    - 0.5
    - 0.5
  dataloader:
    persistent_workers: true
    batch_size: ${train_batch_size}
    num_workers: 20
    shuffle: true
opts: null
args:
  config_file: configs/autoencoder/flexTokenizer/ablation_scaling/vae_gan_scaling_hpo.yaml
  auto_resume: false
  checkpoint_path: null
  opts:
  - opts
  - train.num_train_steps=50000
  - logger.checkpoint_every_step=5000
  - evaluation.eval_for_steps=97
  - evaluation.eval_every_step=2500
  - train_batch_size=4
  - optim.base_lr=2e-4
  - disc_optim.base_lr=2e-4
  - logger.wandb.project=any_tokenizer_official
  - logger.wandb.mode=online
  - model.quantizer_config.use_l2_norm=True
  - model.embedding_dim=16
  - model.quantizer_config.embed_dim=16
  - model.quantizer_config.num_embed=262144
  - model.channel_multipliers=[1,1,2,2,4]
  - model.encoder_layer_configs=[2,2,2,2,2,2]
  - model.decoder_layer_configs=[2,2,2,2,2,2]
  - model.quantizer_config.quantize_type=gvq
  - model.quantizer_config.num_group=8
  - model.quantizer_config.use_l2_norm=false
  output_dir: ./logs_offical_run_tokenizer/ablation_size/GVQ-D-GAN-256-f16-d16g8-v262144-l2-BS256-1_1_2_2_4-no_l2
```

Change data path to where you place hugging face ImageNet.
And change GPU per batch size to have 512 in total.

By the way, how long does your training take?

**Comment by qw85639229** [comment]
Thank you very much for providing the information. Currently, we have set the batch size to 16, using a total of 16 GPUs for training. We plan to train for a total of 100,000 steps, and we are approximately halfway through the training process.

**Comment by qw85639229** [comment]
Hello,

We ran the configuration on 64 cards with a total batch size of 512, but the results are different from what the paper claims. Could you please suggest possible reasons for this discrepancy?
```
seed: 42
dataset: imagenet-1k
train_batch_size: 8

accelerate:
  mixed_precision: bf16

ema:
  beta: 0.999
  
model:
  name: flexTokenizer
  embedding_dim: 16
  hidden_channels: 128
  channel_multipliers: [1, 1, 2, 2, 4]
  encoder_layer_configs: [2, 2, 2, 2, 2, 2]
  decoder_layer_configs: [2, 2, 2, 2, 2, 2]
  use_adaptive_norm: true
  use_learnable_up_down_sample: true
  quantizer_config:
    quantize_type: gvq
    embed_dim: 16
    num_embed: 262144
    commitment_loss_weight: 0.25
    use_l2_norm: false
    use_uniform_init: false
    num_group: 8

discriminator:
  name: dino_discriminator
  lambda_perceptual_loss: 1.0
  disable_adaptive_weight: true
  gen_loss_type: non_saturate
  disc_loss_type: hinge
  lambda_adversarial_loss : 0.1
  lambda_disc_loss: 1.0

  
logger:
  dataset_name: ${dataset} # this is used to apply the correct normalization to the images
  checkpoint_every_step: 5000
  wandb:
    project: any_tokenizer_official
    mode: offline

  image_logger:
    img_log_iter_frequency: 1
    max_images_to_log: 8
    save_on_local: true
    save_on_wandb: true
    rescale_mean: [0.5, 0.5, 0.5]
    rescale_std: [0.5, 0.5, 0.5]
  
train:
  num_train_steps : 50000
  grad_accum_every: 1
  apply_gradient_penalty_every: 0
  max_grad_norm: 2
  discr_start_after_step: 0
  train_batch_size: ${train_batch_size}
  use_auto_expanding: False
  start_expanding_step: 0
  expanding_every_step: 100
  expanding_num: 500

evaluation:
  eval_every_step: 2500
  eval_for_steps: 97
  dataset_name: ${dataset}
  metrics:
    - mse
    - fid
    - is
    # - sfid
    # - fdd
    - lpips
    - psnr
    - ssim

optim:
  base_lr: 2e-4
  lr_scale: fixed
  wd: 0.05
  betas: [0.9, 0.99]
  eps: 1e-8
  
disc_optim:
  base_lr: 2e-4
  lr_scale: fixed
  wd: 0.05
  betas: [0.9, 0.99]
  eps: 1e-8

lr_scheduler:
  name: constant

disc_lr_scheduler:
  name: constant

train_data:
  dataset:
    path: ILSVRC___imagenet-1k
    split: train
  transforms:
    # before normlization, the images should be in range of [0, 1]
    no_aug: True # Do not use any Timm's augmentations for INET training
    is_training: True
    input_size: 256
    mean: [0.5, 0.5, 0.5]
    std: [0.5, 0.5, 0.5]
  dataloader:
    persistent_workers: true
    batch_size: ${train_batch_size}
    num_workers: 20
    shuffle: True

eval_data: 
  dataset:
    path: ILSVRC___imagenet-1k
    split: validation
  transforms:
    no_aug: True # Do not use any Timm's augmentations for INET training
    is_training: False
    input_size: 256
    mean: [0.5, 0.5, 0.5]
    std: [0.5, 0.5, 0.5]
  dataloader:
    persistent_workers: true
    batch_size: ${train_batch_size} 
    num_workers: 20
    shuffle: True
```
The corresponding output is as follows:
```
Directory: gvq_origin
  Step: 50000
  fid: 0.63671875
  is_mean: 210.0
  lpips: 0.07046561688184738
  psnr: 23.392610549926758
  ssim: 0.7035001516342163
  used_codebook: 46.83990478515625
  perplexity: 20810.05859375
```

**Comment by rakkit** [comment]
Did you modify the training loop? 

Use eval code can give a better rFID, the actual sample used in val (with training) is only about 97*512=49664.  

Here are intermedia results (GVQ-D-GAN-256-f16-d16g8-v262144-l2-BS256-1_1_2_2_4-no_l2) for aligning. 
(Original code should only save EMA model's weights)

Testing on eval_10000.pt |  fid:1.4688|perplexity:20917.2188|used_codebook:33.8078|

Testing on eval_20000.pt |fid:0.8008|perplexity:24001.8906|used_codebook:39.6038|

Testing on eval_30000.pt |fid:0.6328|perplexity:22958.4414|used_codebook:41.1449|

Testing on eval_40000.pt |fid:0.5547|perplexity:21949.6953|used_codebook:43.2808|

Testing on eval_50000.pt |fid:0.5078|perplexity:22181.1914|used_codebook:46.8361|

<img width="571" alt="image" src="https://github.com/user-attachments/assets/178dbfb6-8e45-4c53-9e7e-6ca2b71cc757" />
<img width="862" alt="image" src="https://github.com/user-attachments/assets/d9850501-4106-4a04-b438-ab267f9524f2" />
<img width="860" alt="image" src="https://github.com/user-attachments/assets/35351237-f6e4-4f47-9f9f-ba53dc88a4e5" />
